### PR TITLE
Sender for kinesis

### LINKED ISSUE #16: Amazon Kinesis
Amazon Kinesis is a service similar to Kafka. It suggests lower message costs and higher amount of data per message (1MB just like Kafka). It persists all data for minimum 24hrs which means you have to pay certain costs to keep the stream alive.

I've not heard anyone request this specifically, just noting some things as we go along.

### PR DESCRIPTION
**Pull request by devinsba**
Not a huge user of autovalue so I pretty much copied the impl from `sender-sqs`. Please comment if there are issues. IntelliJ doesn't like it but it compiles, so..

Closes #16

### CODE REVIEW

**Comment by adriancole** [comment]
ok base test is working (needed to cbor decode the payload) you can add any more you see fit now. once readme etc polish is in, let's merge time!

File: sender-kinesis/src/test/java/zipkin/reporter/kinesis/KinesisSenderTest.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+import zipkin.reporter.internal.AwaitableCallback;
+
+import static java.util.stream.Collectors.toList;
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class KinesisSenderTest {
+  @Rule public MockWebServer server = new MockWebServer();
+
+  // Kinesis sends data in CBOR format
+  ObjectMapper mapper = new ObjectMapper(new CBORFactory());
```
Comment (adriancole) [inline sender-kinesis/src/test/java/zipkin/reporter/kinesis/KinesisSenderTest.java]: first I've ever seen this.. neat! https://github.com/FasterXML/jackson-dataformats-binary/tree/master/cbor

File: sender-kinesis/src/main/java/zipkin/reporter/kinesis/KinesisSender.java [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+  // Needed to be able to overwrite for tests
+  @Nullable
+  abstract AwsClientBuilder.EndpointConfiguration endpointConfiguration();
+
+  private final AtomicBoolean closeCalled = new AtomicBoolean(false);
+
+  @Override
+  public CheckResult check() {
+    try {
+      // TODO are we OK requiring DescribeStream be available to us as a publisher?
```
Comment (adriancole) [inline sender-kinesis/src/main/java/zipkin/reporter/kinesis/KinesisSender.java]: I'm ok with it. We can add to the readme, and if someone complains, we can always address later.

ps this should be mock testable (the success and failure case)

**Comment by devinsba** [comment]
Thanks for tracking it down. I've never even heard of CBOR before now

File: sender-kinesis/pom.xml [hunk 3]
```diff
@@ -54,6 +57,42 @@
       <version>3.7.0</version>
       <scope>test</scope>
     </dependency>
+
+    <dependency>
```
Comment (adriancole) [inline sender-kinesis/pom.xml]: not sure you want to compile scope these, right?

File: sender-kinesis/src/test/java/zipkin/reporter/kinesis/KinesisSenderTest.java [hunk 4]
```diff
@@ -64,6 +65,39 @@ public void sendsSpans() throws Exception {
...
+  //  assertThat(result.exception).isInstanceOf(IllegalStateException.class);
+  //}
+
+  @Test
+  public void checkFailsWithException() {
+    server.enqueue(new MockResponse());
+
+    Component.CheckResult result = sender.check();
+    assertThat(result.ok).isFalse();
+    assertThat(result.exception).isInstanceOf(NullPointerException.class);
```
Comment (adriancole) [inline sender-kinesis/src/test/java/zipkin/reporter/kinesis/KinesisSenderTest.java]: hehe

Comment (adriancole) [inline sender-kinesis/src/test/java/zipkin/reporter/kinesis/KinesisSenderTest.java]: Unless you want to make sure this is an NPE, you can probably drop this assert

Comment (devinsba) [inline sender-kinesis/src/test/java/zipkin/reporter/kinesis/KinesisSenderTest.java]: I'd like to find a way to force it to end up with an AWS exception or Apache http client exception so I can actually see the exception from the catch block

**Comment by devinsba** [comment]
That approach is certainly much nicer. Thanks for sharing it :+1:

**Pull request merged**